### PR TITLE
[Publisher] Admin Panel: Agency Provisioning - Create/Edit Agency Information (4/n)

### PR DESCRIPTION
**Pull request by mxosman**
## Description of the change

Implements the create/edit agency modal for the Agency Provisioning flow. There are two tabs in this modal - 
1. Agency Information - allows you to edit information related to the agency
2. Team Members - allows you to add, remove, and/or assign roles to users connected to this agency (will be implemented in the next PR)

Demo (creating new/editing existing agency):

https://github.com/Recidiviz/justice-counts/assets/59492998/21a0e40f-1c5e-4ad7-b5bf-cc238835f905

https://github.com/Recidiviz/justice-counts/assets/59492998/25ea6bb6-1444-42ee-894e-2c6aef988962

## Related issues

Contributes to #1051

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [x] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx [hunk 1]
```diff
@@ -15,10 +15,737 @@
...
+          currentSettingType === AgencyProvisioningSettings.TEAM_MEMBERS_ROLES,
+      },
+    ];
+
+    /** Selected agency to edit */
+    const selectedAgency = selectedIDToEdit
+      ? agenciesByID[selectedIDToEdit][0]
+      : undefined;
+
+    /** Available agencies ("available" meaning excluding the current agency) to select from */
```
Comment (lilidworkin) [inline publisher/src/components/AdminPanel/AgencyProvisioning.tsx:127]: This was a helpful comment!

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx [hunk 2]
```diff
@@ -15,10 +15,737 @@
...
+      ];
+    };
+
+    const saveUpdates = async () => {
+      setIsSaveInProgress(true);
+
+      /** Update final list of systems, child agencies, and team members */
+      updateSystems(Array.from(selectedSystems));
+      updateChildAgencyIDs(Array.from(selectedChildAgencyIDs));
+      updateTeamMembers([]);
```
Comment (lilidworkin) [inline publisher/src/components/AdminPanel/AgencyProvisioning.tsx:183]: is team member supposed to be empty?

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx [hunk 3]
```diff
@@ -15,10 +15,737 @@
...
+
+    /**
+     * Check whether user has made updates to various fields to determine whether or not the 'Save' button is enabled/disabled
+     *
+     * Note: when creating a new agency, the only required fields are the `name` and `state_code`, all other fields can be
+     *       updated at a later time.
+     * */
+
+    /**
+     * Existing agency: an update has been made when the agency has a value for `agencyProvisioningUpdates.name`
```
Comment (lilidworkin) [inline publisher/src/components/AdminPanel/AgencyProvisioning.tsx:235]: Great comments!

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx [hunk 4]
```diff
@@ -15,10 +15,737 @@
...
+      ? agenciesByID[selectedIDToEdit][0]
+      : undefined;
+
+    /** Available agencies ("available" meaning excluding the current agency) to select from */
+    const agencyIDs = agencies.map((agency) => +agency.id);
+    const availableAgencies = agencies.filter(
+      (agency) => agency.id !== selectedAgency?.id
+    );
+
+    /** A list of superagencies and child agencies to select from */
```
[the remark this answers is not in the document]
Comment (lilidworkin) [reply publisher/src/components/AdminPanel/AgencyProvisioning.tsx:133]: Or maybe we don't even need to change the definition of available agencies?! If selectedAgency is null will available agencies == agencies?

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx [hunk 5]
```diff
@@ -278,8 +278,8 @@ export const AgencyProvisioning: React.FC<ProvisioningProps> = observer(
      * boolean flag for that property before the modal was open.
      */
     const hasDashboardEnabledStatusUpdate =
-      agencyProvisioningUpdates.is_dashboard_enabled !==
-      selectedAgency?.is_dashboard_enabled;
+      Boolean(agencyProvisioningUpdates.is_dashboard_enabled) !==
```
Comment (lilidworkin) [inline publisher/src/components/AdminPanel/AgencyProvisioning.tsx:281]: curious why is this needed?

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx (hunk 2 above)
Comment (lilidworkin) [reply publisher/src/components/AdminPanel/AgencyProvisioning.tsx:183]: Oh got it :D

File: publisher/src/components/AdminPanel/AgencyProvisioning.tsx (hunk 5 above)
Comment (mxosman) [reply publisher/src/components/AdminPanel/AgencyProvisioning.tsx:281]: Great question! I was playing around and ran into this funky behavior. This is to hedge against the cases where `is_dashboard_enabled` comes up `null`. We may never run into that case since is_dashboard_enabled` will be enabled by default, but if the FE receives a `null` for this property for an agency and you click into that agency -
* You'll see that the `Dashboard enabled` checkbox is unchecked - saving is disabled because there are no changes
* If you click the checkbox to mark it checked - saving is enabled because it detects a change
* If you click the checkbox one more time to uncheck it - saving is still enabled because it was previously `null` and the unchecking it made it `false` - and so the updater logic will evaluate `null !== false` and return `true` for `hasDashboardEnabledStatusUpdate`

So coercing these to a boolean and evaluating them felt like the safest way to get an accurate idea of whether or not someone made an update to this field.


https://github.com/Recidiviz/justice-counts/assets/59492998/4735f3b3-d2bf-4928-a0fd-55374148a03d

**Comment by mxosman** [comment]
> Not my most thorough of reviews but I think this looks good :D

I'll take any reviews! They're always super helpful! TYSM, Lili!

**Comment by mxosman** [comment]
There's still some bug bashing to do for this flow and I remembered that I need to orchestrate the adding and removing of the Superagency system as users toggle that checkmark on/off -- or when they add a Superagency system, to trigger that checkbox to turn on, etc. I'll work on these items in a follow up PR after the team members PR!
